ros/src/computing/perception/localization/packages/ekf_localizer/src/ekf_localizer.cpp

```diff
@@ -0,0 +1,672 @@
+  if (delay_step > extend_state_step_ - 1)
+  {
+    ROS_WARN_DELAYED_THROTTLE(1.0, "Twist delay exceeds the compensation limit, ignored. delay: %f[s], limit = extend_state_step * ekf_dt : %f [s]",
+                              delay_time, extend_state_step_ * ekf_dt_);
+    return;
+  }
+  DEBUG_INFO("delay_time: %f [s]", delay_time);
+
+  /* Set measurement matrix */
+  Eigen::MatrixXd y(dim_y, 1);
+  y << twist.twist.linear.x, twist.twist.angular.z;
+
+  /* Gate */
+  Eigen::MatrixXd y_ekf(dim_y, 1);
+  y_ekf << ekf_.getXelement(delay_step * dim_x_ + IDX::VX), ekf_.getXelement(delay_step * dim_x_ + IDX::WZ);
+  Eigen::MatrixXd P_curr;
+  ekf_.getLatestP(P_curr);
+  if (!mahalanobisGate(pose_gate_dist_, y_ekf, y, P_curr.block(4, 4, dim_y, dim_y)))
+  {
+    ROS_WARN_DELAYED_THROTTLE(2.0, "[EKF] Twist measurement update, mahalanobis distance is over limit. ignore measurement data.");
+    return;
+  }
+
+  DEBUG_PRINT_MAT(y.transpose());
+  DEBUG_PRINT_MAT(y_ekf.transpose());
+  DEBUG_PRINT_MAT((y - y_ekf).transpose());
+
+  /* Set measurement matrix */
+  Eigen::MatrixXd C = Eigen::MatrixXd::Zero(dim_y, dim_x_);
+  C(0, IDX::VX) = 1.0; // for vx
+  C(1, IDX::WZ) = 1.0; // for wz
+
+  /* Set measurement noise covariancs */
+  Eigen::MatrixXd R = Eigen::MatrixXd::Zero(dim_y, dim_y);
+  R(0, 0) = twist_stddev_vx_ * twist_stddev_vx_ * ekf_dt_; // for vx
+  R(1, 1) = twist_stddev_wz_ * twist_stddev_wz_ * ekf_dt_; // for wz
+
+  /* In order to avoid a large change by update, measurement update is performed by dividing at every step. */
+  R *= (ekf_rate_ / twist_rate_);
+
+  ekf_.updateWithDelay(y, C, R, delay_step);
+
+  // debug
+  Eigen::MatrixXd X_result(dim_x_, 1);
+  ekf_.getLatestX(X_result);
+  DEBUG_PRINT_MAT(X_result.transpose());
+  DEBUG_PRINT_MAT((X_result - X_curr).transpose());
+};
+
+/*
+ * mahalanobisGate
+ */
+bool EKFLocalizer::mahalanobisGate(const double &dist_max, const Eigen::MatrixXd &x,
+                                       const Eigen::MatrixXd &obj_x, const Eigen::MatrixXd &cov)
+{
+  Eigen::MatrixXd mahalanobis_squared = (x - obj_x).transpose() * cov.inverse() * (x - obj_x);
+  DEBUG_INFO("measurement update: mahalanobis = %f, gate limit = %f", std::sqrt(mahalanobis_squared(0)), dist_max);
+  if (mahalanobis_squared(0) > dist_max * dist_max)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * publishEstimatedPose
+ */
+void EKFLocalizer::publishEstimatedPose()
+{
+  ros::Time current_time = ros::Time::now();
+  Eigen::MatrixXd X(dim_x_, 1);
+  Eigen::MatrixXd P(dim_x_, dim_x_);
+  ekf_.getLatestX(X);
+  ekf_.getLatestP(P);
+
+  /* publish latest pose */
+  pub_pose_.publish(current_ekf_pose_);
+
+  /* publish latest pose with covariance */
+  geometry_msgs::PoseWithCovarianceStamped pose_cov;
+  pose_cov.header.stamp = current_time;
+  pose_cov.header.frame_id = current_ekf_pose_.header.frame_id;
+  pose_cov.pose.pose = current_ekf_pose_.pose;
+  for (int i = 0; i < 36; ++i)
```

Comment:
this is [unnecessary](http://wiki.ros.org/roscpp/Overview/Messages#Message_Generation).

Reply:
Sorry, which part did you mention?

Reply:
The initialization of the covariance matrix. 
The for block contained in
`for (int i = 0; i < 36; ++i)`

Reply:
> Note that all fields in a message are currently initialized to a default value of 0 (or empty for strings and variable-length arrays).

I get it. thanks!

Reply:
remove initialization. pls check it.